homeassistant/components/xiaomi_aqara/config_flow.py

```diff
@@ -0,0 +1,183 @@
+                        return await self.async_step_select()
+
+                    errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=GATEWAY_CONFIG, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple aqara gateways found."""
+        errors = {}
+        if user_input is not None:
+            ip_adress = user_input["select_ip"]
+            self.selected_gateway = self.gateways[ip_adress]
+            return await self.async_step_settings()
+
+        select_schema = vol.Schema(
+            {
+                vol.Required("select_ip"): vol.In(
+                    [gateway.ip_adress for gateway in self.gateways.values()]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_schema, errors=errors
+        )
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle zeroconf discovery."""
+        name = discovery_info.get("name")
+        self.host = discovery_info.get("host")
+        mac_address = discovery_info.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # Check if the discovered device is an xiaomi aqara gateway.
+        if not name.startswith(ZEROCONF_GATEWAY):
+            _LOGGER.debug(
+                "Xiaomi device '%s' discovered with host %s, not identified as xiaomi aqara gateway",
+                name,
+                self.host,
+            )
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # format mac (include semicolns and make uppercase)
+        mac_address = format_mac(mac_address)
+
+        unique_id = mac_address
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured({CONF_HOST: self.host})
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update({"title_placeholders": {"name": self.host}})
+
+        return await self.async_step_user()
+
+    async def async_step_settings(self, user_input=None):
+        """Specify settings and connect aqara gateway."""
+        errors = {}
+        if user_input is not None:
+            # get all required data
+            name = user_input[CONF_NAME]
+            key = user_input.get(CONF_KEY)
+            ip_adress = self.selected_gateway.ip_adress
+            port = self.selected_gateway.port
+            sid = self.selected_gateway.sid
+            protocol = self.selected_gateway.proto
+
+            if key is not None:
+                # validate key by issueing stop ringtone playback command.
+                self.selected_gateway.key = key
+                valid_key = self.selected_gateway.write_to_hub(sid, mid=10000)
+            else:
+                valid_key = True
```

Comment:
What is the use case for not providing a key?

Reply:
The current integration also offers the capability of not providing a key.
I guess it is for people that can't get the key....
I know some gateway models are not able to retrieve the key (although I think with those gateways the whole port is closed so you also can't connect to the gateway for the sensors).

However I am not sure, why you would not specify a key...
I just did not want to break it for people that use the current integration without a key.

Reply:
Without a key you can get the state of all devices and sensors, you just can't control any device. But If you only own sensors it doesn't really matter...

Reply:
Should we then limit the platforms to only sensor platforms when there's no key set? The other platforms won't work without a key.

Reply:
Good sugestion, I limited it to only sensors and binary_sensors if no key is provided